File: 3-months-prep/03-05-migratory-birds.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.026;
use Carp::Assert;


sub migratoryBirds {
	my @arr = @{$_[0]};
	my $result_value = -2**64+1;
	my $result_index = undef;
	my %counts;
	foreach my $x (@arr) {
		$counts{$x} += 1;
	}
	foreach my $k (sort {$a<=>$b} keys %counts) {
		my $v = $counts{$k};
		if ($v > $result_value) {
			$result_index = $k;
			$result_value = $v;
		}
	}
	return $result_index;
}

my @input_values = ( [1,4,4,4,5,3], [1,2,3,4,5,4,3,2,1,3,4], );
my @input_checks = ( 4, 3, );
assert (scalar(@input_values) == scalar(@input_checks));

for (my $i = 0; $i < scalar(@input_values); ++$i) {
	my @arr = @{$input_values[$i]};
	my $check = $input_checks[$i];
	say "arr=(@arr)";
	my $result = migratoryBirds \@arr;
	say "result=($result)";
	assert ($result == $check);
	say "";
}

File: 3-months-prep/03-05-migratory-birds.py
import math
from collections import Counter
from typing import List

def migratoryBirds(arr):
    result_value = -math.inf
    result_index = None
    counts = Counter(arr)
    #for k, v in counts.items():
    for k in sorted(counts.keys()):
        v = counts[k]
        if v > result_value:
            result_index = k
            result_value = v
    return result_index


input_values = [ [1,4,4,4,5,3], [1,2,3,4,5,4,3,2,1,3,4], ]
input_checks = [ 4, 3, ]
assert (len(input_values) == len(input_checks))

for arr, check in zip(input_values, input_checks):
    print("arr=(%s)" % arr)
    result = migratoryBirds(arr)
    print("result=(%s)" % result)
    assert (result == check), "Check comparison failed"
    print()

